src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp

```diff
@@ -46,6 +46,102 @@
 
 typedef void (MacroAssembler::* chr_insn)(Register Rt, const Address &adr);
 
+// jdk.internal.util.ArraysSupport.vectorizedHashCode
+address C2_MacroAssembler::arrays_hashcode(Register ary, Register cnt, Register result,
+                                           BasicType eltype) {
+  assert_different_registers(ary, cnt, result, rscratch1, rscratch2);
+
+  Register tmp1 = rscratch1, tmp2 = rscratch2;
+
+  Label TAIL, RELATIVE;
+
+  const size_t unroll_factor = 4;
+  const size_t loop_factor = eltype == T_BOOLEAN || eltype == T_BYTE ? 32
+                             : eltype == T_CHAR || eltype == T_SHORT ? 16
+                             : eltype == T_INT                       ? 16
+                                                                     : 0;
+  guarantee(loop_factor, "unsupported eltype");
+
+  switch (eltype) {
+  case T_BOOLEAN:
+    BLOCK_COMMENT("arrays_hashcode(unsigned byte) {");
+    break;
+  case T_CHAR:
+    BLOCK_COMMENT("arrays_hashcode(char) {");
+    break;
+  case T_BYTE:
+    BLOCK_COMMENT("arrays_hashcode(byte) {");
+    break;
+  case T_SHORT:
+    BLOCK_COMMENT("arrays_hashcode(short) {");
+    break;
+  case T_INT:
+    BLOCK_COMMENT("arrays_hashcode(int) {");
+    break;
+  default:
+    ShouldNotReachHere();
+  }
+
+  subsw(cnt, cnt, loop_factor);
+  br(Assembler::LO, TAIL);
+
+  RuntimeAddress stub = RuntimeAddress(StubRoutines::aarch64::large_arrays_hashcode(eltype));
+  assert(stub.target() != nullptr, "array_hashcode stub has not been generated");
+  address tpc = trampoline_call(stub);
+  if (tpc == nullptr) {
+    DEBUG_ONLY(reset_labels(TAIL, RELATIVE));
+    postcond(pc() == badAddress);
+    return nullptr;
+  }
+
+  bind(TAIL);
+
+  assert(is_power_of_2(unroll_factor), "can't use this value to calculate the jump target PC");
+  orr(tmp2, cnt, 0x1fff ^ (unroll_factor - 1));
```

Comment:
This is rather cryptic, especially the slightly arbitrary choice of 0x1fff -- any all 1s value greater than (loop_factor - 1) would do. I think a comment as to what is being done here might help maintainers understand what is happening here.

// At this point cnt holds (r - l) where r is the number of remaining elements,
// l is loop_count and 0 <= r < l. The constant operand to the orr 'rounds' this
// negative value into range [-u, -1] where u = unroll_factor, by clearing bits
// below bit k = log2(unroll_factor) and setting any higher bits that might be clear.
// The subtract shifted by 3 offsets past ((u - r) % u) load + madd insns i.e. it only
// executes r % u load + madds. Iteration eats up the remainder, u elements at a time.

Reply:
Thank you for providing a detailed comment!

I can't remember why I needed the `0x1fff` literal. Probably this is an artifact left from WIP code. I believe this is not required anymore and the `orr` can be simplified to `orr(tmp2, cnt, -unroll_factor)`.

Given that, I'd suggest to shorten the comment to:

```c++
  // At this point cnt holds (r - l) where r is the number of remaining elements, l is loop_count                                                                                                                                                                                                                           
  // and 0 <= r < l. The orr performs (r - l) % u where u = unroll_factor.  The subtract shifted by                                                                                                                                                                                                                         
  // 3 offsets past |(r - l) % u| load + madd insns i.e. it only executes r % u load + madds.                                                                                                                                                                                                                               
  // Iteration eats up the remainder, u elements at a time.
```

Reply:
Yes, just fixing the constant helps a lot. However, the comment is still worth having.